### PR TITLE
MANTA-4341 Optimize conditional object operations

### PR DESCRIPTION
**Pull request by chudley**
These are the tests I've been using the check my work on MANTA-4341. It's not totally clean yet, and I'm still not sure about things like the lack of `res.statusCode` on a PreconditionFailedError.

I wanted to wrap it up into a PR though because the other parts of MANTA-4341 are due to be reviewed and it'll be good to at least prove there's been _some_ testing. Perhaps it'll also be good to get some early review as well to make sure it's not a million miles off.

### CODE REVIEW

**Comment by chudley** [comment]
Ah, looks like that property is only available from Node v13 (https://github.com/nodejs/node/pull/28814).

I'll see if there's a better way to verify this behaviour in the test suite.

**Comment by chudley** [comment]
So we can't get at these properties in earlier versions of node unless we reach into `_readableState` ourselves. It turns out [lots of people do it anyway](https://github.com/nodejs/node/issues/445).

I haven't put a great deal of thought into these tests so there's a good chance they're not testing the right thing anyway. I've removed them in this latest push, but I'd be happy to revisit if you think otherwise!

**Comment by kellymclaughlin** [comment]
> I haven't put a great deal of thought into these tests so there's a good chance they're not testing the right thing anyway. I've removed them in this latest push, but I'd be happy to revisit if you think otherwise!

I *think* it's probably ok to leave them out versus reaching into `_readableState`. I feel like the other assertions should give us confidence that the operations are working as intended so hopefully it's ok without these checks.

File: test/integration/buckets-client-condreq.test.js [hunk 1]
```diff
@@ -95,21 +108,44 @@ test('buckets client conditional requests', testOpts, function (suite) {
...
                 headers: {
-                    'if-modified-since': 'XXX'
+                    'if-unmodified-since': 'x'
+                }
+            },
+            function (err, res) {
+                t.ok(err);
+                t.ok(res);
+                t.equal(res.statusCode, 400);
+                //console.log(err);
```
Comment (kellymclaughlin) [inline test/integration/buckets-client-condreq.test.js:135]: Can this be removed?

File: test/integration/buckets-client-condreq.test.js [hunk 2]
```diff
@@ -118,66 +154,375 @@ test('buckets client conditional requests', testOpts, function (suite) {
...
+        client.headBucketObject(
+            BUCKET_NAME,
+            OBJECT_NAME,
+            {
+                headers: {
+                    'if-unmodified-since': d
+                }
+            },
+            function (err, res) {
+                //t.ok(err);
```
Comment (kellymclaughlin) [inline test/integration/buckets-client-condreq.test.js:214]: Remove?

File: test/integration/buckets-client-condreq.test.js [hunk 3]
```diff
@@ -118,66 +154,375 @@ test('buckets client conditional requests', testOpts, function (suite) {
...
+        client.putBucketObjectMetadata(BUCKET_NAME, OBJECT_NAME, {
+            headers: {
+                'm-foo': 'test',
+                'if-match': libuuid.v4()
+            } }, function (err, res) {
+
+            t.ok(err);
+            t.ok(res);
+            /*
+             * XXX Why no `res.statusCode` from client, like `get` does.
```
Comment (kellymclaughlin) [inline test/integration/buckets-client-condreq.test.js:491]: This is really odd. Wonder why it would be different in this case?

Comment (chudley) [reply test/integration/buckets-client-condreq.test.js:486]: Oh man, sorry about this one. I think my brain is a bit frazzled. I just spent way too long looking at why node-manta wasn't parsing a response body from a HEAD request...

So the difference here is that node-manta's buckets client does a lot of the response wrangling itself, but for GETs it leans on the existing directory client for some niceties around resumability.

The difference is that the buckets error path doesn't return `res` on error and instead populates `err` with all the information. Directories is different where `res` is included as well it seems, so this is why GET in the buckets client also includes `res`.

The latest commit should account for this in the test suite. I think there is still some work to do to make this consistent across all client requests, but it's not a million miles off and I expect explains why the buckets work is still on its own branch.

**Pull request merged**